lib/gateway/build-gateway.js

```diff
@@ -508,12 +512,29 @@ async function buildGateway (serviceMap, gatewayOpts, app, lruGatewayResolvers)
     serviceMap,
     entityResolversFactory: factory,
     pollingInterval: gatewayOpts.pollingInterval,
+    serviceFn: typeof gatewayOpts.services === 'function' ? gatewayOpts.services : undefined,
     async refresh (isRetry) {
       const failedMandatoryServices = []
       if (this._serviceSDLs === undefined) {
         this._serviceSDLs = serviceSDLs.join(' ')
       }
 
+      if (this.serviceFn) {
+        const newServices = await this.serviceFn()
+        const oldServices = Object.keys(serviceMap)
+        const addedServices = newServices.filter(({ name }) => !oldServices.includes(name))
+        const deletedServices = oldServices.filter(name => !newServices.find(service => service.name === name))
+        for (const name of deletedServices) {
+          serviceMap[name]
+            .close()
+            .catch(
+              // istanbul ignore next
```

Comment:
are we able to get some coverage here to check the behaviour?

Reply:
Maybe I can force an error in the close function and check that nothing happens?

`test('should not throw when an error happens on the closing function')`

and then have something like `serviceMap[postService].close = () => throw new Error()`

WDYT?

Reply:
yeah perfect!

Reply:
@jonnydgreen I think it should all be ok now